### DISCUSSION TITLE
Installation on non default port

### DISCUSSION THREAD

**Discussion by ckorder** [searx/searx#2381] [Installation issue: Official wiki]
https://searx.github.io/searx/admin/installation.html
https://searx.github.io/searx/admin/installation-searx.html
https://searx.github.io/searx/admin/installation-nginx.html

please make it easy for use different ports because 3000 and 8888 if often taken by other services. ( many docker container, jitsi meet etc.)
i changed the ports in:
./utils/searx.sh
./utils/filtron.sh
./utils/morty.sh
./searx/settings.yml

but still get this:

 **|searx| DEBUG:searx.webapp:starting webserver on 127.0.0.1:8888**
leads to:  **OSError: [Errno 98] Address already in use**
even when i changed the ports in the configs above to 8889
i changed 3000 to 3001 in the config and the nginx config because both ports conflict my other services. within docker i got another response but it was still not working, so either provide ease port changes in normal or in the docker version, best would be making both portable.

please make a call if i got there something wrong but i was not able fix the issue in my sparetime.

```
curl --location --verbose --head --insecure 127.0.0.1:8889
  |searx| INFO:searx:merge the default settings ( /usr/local/searx/searx-src/searx/settings.yml ) and the user setttings ( /etc/searx/settings.yml )
  |searx| INFO:searx:Initialisation done
  |searx| INFO:searx.plugins.https_rewrite:29 rules loaded
  |searx| INFO:searx.search:max_request_timeout=None
  |searx| DEBUG:searx.webapp:static directory is /usr/local/searx/searx-src/searx/static
  |searx| DEBUG:searx.webapp:templates directory is /usr/local/searx/searx-src/searx/templates
  |searx| DEBUG:searx.webapp:starting webserver on 127.0.0.1:8888
  |searx|  * Serving Flask app "webapp" (lazy loading)
  |searx|  * Environment: production
  |searx|    WARNING: This is a development server. Do not use it in a production deployment.
  |searx|    Use a production WSGI server instead.
  |searx|  * Debug mode: on
  |searx| Traceback (most recent call last):
  |searx|   File "searx/webapp.py", line 1143, in <module>
  |searx|     run()
  |searx|   File "searx/webapp.py", line 1097, in run
  |searx|     threaded=True
  |searx|   File "/usr/local/searx/searx-pyenv/lib/python3.7/site-packages/flask/app.py", line 990, in run
  |searx|     run_simple(host, port, self, **options)
  |searx|   File "/usr/local/searx/searx-pyenv/lib/python3.7/site-packages/werkzeug/serving.py", line 1030, in run_simple
  |searx|     s.bind(server_address)
  |searx| OSError: [Errno 98] Address already in use
  |searx| * Expire in 0 ms for 6 (transfer 0x56259979abf0)
  |searx| *   Trying 127.0.0.1...
  |searx| * TCP_NODELAY set
  |searx| * Expire in 200 ms for 4 (transfer 0x56259979abf0)
  |searx|   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
  |searx|                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 8889 (#0)
  |searx| > HEAD / HTTP/1.1
  |searx| > Host: 127.0.0.1:8889
  |searx| > User-Agent: curl/7.64.0
  |searx| > Accept: */*
  |searx| > 
  |searx| < HTTP/1.1 500 Internal Server Error
  |searx| < Connection: close
  |searx| < Content-Type: text/plain
  |searx| < 
  |searx| * Excess found in a non pipelined read: excess = 21 url = / (zero-length body)
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  |searx| * Closing connection 0
  |searx| HTTP/1.1 500 Internal Server Error
  |searx| Connection: close
  |searx| Content-Type: text/plain
  |searx| 
 
```

nginx config:

> server {
> 
> 
> 
> server_name search.mydomain.com;
> 
> location /searx {
>     proxy_pass         http://127.0.0.1:4004/;
> 
>     proxy_set_header   Host             $http_host;
>     proxy_set_header   Connection       $http_connection;
>     proxy_set_header   X-Real-IP        $remote_addr;
>     proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;
>     proxy_set_header   X-Scheme         $scheme;
>     proxy_set_header   X-Script-Name    /searx;
> }
> 
> location /searx/static {
>     alias /usr/local/searx/searx-src/searx/static;
> }
> 
> 
> location /morty {
>     proxy_pass         http://127.0.0.1:3001/;
> 
>     proxy_set_header   Host             $http_host;
>     proxy_set_header   Connection       $http_connection;
>     proxy_set_header   X-Real-IP        $remote_addr;
>     proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;
>     proxy_set_header   X-Scheme         $scheme;
> }
> 
>     listen 443 ssl; # managed by Certbot
>     ssl_certificate /etc/letsencrypt/live/search.mydomain.org/fullchain.pem; # managed by Certbot
>     ssl_certificate_key /etc/letsencrypt/live/search.mydomain.org/privkey.pem; # managed by Certbot
> 
>     include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
>     ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
> 
> }
> 
> server {
>     if ($host = search.mydomain.com) {
>         return 301 https://$host$request_uri;
>     } # managed by Certbot
> 
> server_name search.mydomain.com;
>     listen 80;
>     return 404; # managed by Certbot
> 
> }

Result in: 
https://search.mydomain.org/searx
Internal Server Error

mydomain.com/morty
502 Bad Gateway
nginx

please allow providing multiple foss services by using non default ports easily

**Comment by return42** [comment]
Do you asking for the [Installation scripts](https://searx.github.io/searx/admin/installation.html#installation-scripts),  did you read chapter [Tooling box setup](https://searx.github.io/searx/utils/index.html#tooling-box-setup)?

update: I added a note to the documentation in PR #2384

**Comment by ckorder** [comment]
I read the installation chapter, and used the Installation scripts. I detected that the docker container later on was still running. I haven't checked any further but i will take another look on it. At all this referred to the non clear and easy change of the used default ports, which could be improved. 

Actually I'm quite pissed. 
settings.yml  
```
server:
    port : 8889
    bind_address : "127.0.0.1" # address to listen on
```

results in:

```
  |searx| DEBUG:searx.webapp:static directory is /usr/local/searx/searx-src/searx/static
  |searx| DEBUG:searx.webapp:templates directory is /usr/local/searx/searx-src/searx/templates
  |searx| DEBUG:searx.webapp:starting webserver on 127.0.0.1:8888
```

searx.sh  
`SEARX_INTERNAL_HTTP="${SEARX_INTERNAL_HTTP:-127.0.0.1:8889}"`

really messy and still not working. 
get your sh1t together instead of pointing towards documentation when its certainly unclear how to run the service on a non default port or make a reasonable call like no time to look into it. Please tell me what about the issue you do not understand, beside annotate the fact that the documentation towards running on a non default port is missing and the process looks like way more than changing these 2 files.

sorry i have to call it:
```

ERROR: missing command lxc

LXD is needed, to install run::

  snap install lxd
  lxd init --auto
```
```
who the hell uses snap crap?

$ sudo -H ./utils/lxc.sh build

A cup of coffee later, your LXC suite is build up and you can run whatever task you want / in a selected or even in all LXC suite containers. If you do not want to build all containers, you can build just one:

$ sudo -H ./utils/lxc.sh build searx-ubu1804

just waste the time of your customers np, no donations from me so far incoming!
```

No one has the time to waste hours of reading documentation subpages without knowing where it goes...
Please make a dumb ass documentation for nginx, easy and strict to follow, best would be an all in one installer with the possibility to change the default ports. Otherwise many people wont use this service, which is quite sad because its good.
The documentation and the installation process for non default port seems to be trash, but maybe I'm wrong, I hope I'm!

```

If needed, set PUBLIC_URL of your WEB service in the '.config.sh' file::
  PUBLIC_URL          : http://10.1.0.4/searx
  SEARX_INSTANCE_NAME : searx@10.1.0.4 
  SERVICE_USER        : searx
  SEARX_INTERNAL_HTTP : http://127.0.0.1:8888
```
.config.sh? where?
~/searx/utils$ ls -la | grep .config.sh
~/searx$ ls -la | grep .config.sh
-rw-r--r--  1 bla bla Dec 11 22:35 .config.sh

nice hidden file and hidden in the sub sub documentation for https://searx.github.io/searx/utils/index.html#tooling-box-setup

while it should be called out in: https://searx.github.io/searx/admin/installation.html#installation-scripts
or just use an install script with one default config file and than not million key pressed to get an installation done!!!!!!!!!!!!!!

** press any [KEY] to continue **
please waste more time instead of just calling out error and reason or done, happy searching. 

`export SEARX_SETTINGS_PATH="/etc/searx/settings.yml"
cd /usr/local/searx/searx-src
timeout 10 python searx/webapp.py &
sleep 3
curl --location --verbose --head --insecure 127.0.0.1:8889
  |searx| INFO:searx:merge the default settings ( /usr/local/searx/searx-src/searx/settings.yml ) and the user setttings ( /etc/searx/settings.yml )
  |searx| INFO:searx:Initialisation done
  |searx| INFO:searx.plugins.https_rewrite:29 rules loaded
  |searx| INFO:searx.search:max_request_timeout=None
  |searx| DEBUG:searx.webapp:static directory is /usr/local/searx/searx-src/searx/static
  |searx| DEBUG:searx.webapp:templates directory is /usr/local/searx/searx-src/searx/templates
  |searx| DEBUG:searx.webapp:starting webserver on 127.0.0.1:8888
  |searx|  * Serving Flask app "webapp" (lazy loading)
  |searx|  * Environment: production
  |searx|    WARNING: This is a development server. Do not use it in a production deployment.
  |searx|    Use a production WSGI server instead.
  |searx|  * Debug mode: on
  |searx| Traceback (most recent call last):
  |searx|   File "searx/webapp.py", line 1143, in <module>
  |searx|     run()
  |searx|   File "searx/webapp.py", line 1097, in run
  |searx|     threaded=True
  |searx|   File "/usr/local/searx/searx-pyenv/lib/python3.7/site-packages/flask/app.py", line 990, in run
  |searx|     run_simple(host, port, self, **options)
  |searx|   File "/usr/local/searx/searx-pyenv/lib/python3.7/site-packages/werkzeug/serving.py", line 1030, in run_simple
  |searx|     s.bind(server_address)
  |searx| OSError: [Errno 98] Address already in use
  |searx| * Expire in 0 ms for 6 (transfer 0x558b6f141bf0)
  |searx| *   Trying 127.0.0.1...
  |searx| * TCP_NODELAY set
  |searx| * Expire in 200 ms for 4 (transfer 0x558b6f141bf0)
  |searx|   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
  |searx|                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* connect to 127.0.0.1 port 8889 failed: Connection refused
  |searx| * Failed to connect to 127.0.0.1 port 8889: Connection refused
  |searx| * Closing connection 0
  |searx| curl: (7) Failed to connect to 127.0.0.1 port 8889: Connection refused
`

just change the 4th file beside the settings secret key in:
 
 cat webapp.py | grep 8888   
cat webutils.py | grep 8888
......
i dont get it what the hell u done with this configs but i looks like more than ...
** press any [KEY] to confirm you are wasting your time **
** press any [KEY] to confirm you are wasting your time **
** press any [KEY] to confirm you are wasting your time **
** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **** press any [KEY] to confirm you are wasting your time **


Dec 22 05:06:41  nginx[14557]: nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
Dec 22 05:06:41  nginx[14557]: nginx: [emerg] bind() to [::]:80 failed (98: Address already in use)
Dec 22 05:06:42  nginx[14557]: nginx: [emerg] bind() to [::]:443 failed (98: Address already in use)
Dec 22 05:06:42  nginx[14557]: nginx: [emerg] bind() to 0.0.0.0:443 failed (98: Address already in use)
Dec 22 05:06:42  nginx[14557]: nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
Dec 22 05:06:42  nginx[14557]: nginx: [emerg] bind() to [::]:80 failed (98: Address already in use)
Dec 22 05:06:42  nginx[14557]: nginx: [emerg] still could not bind()
Dec 22 05:06:42  systemd[1]: nginx.service: Control process exited, code=exited, status=1/FAILURE
Dec 22 05:06:42  systemd[1]: nginx.service: Failed with result 'exit-code'.
Dec 22 05:06:42  systemd[1]: Failed to start A high performance web server and a reverse proxy server.

sure, looks like nginx runs twice. never seen this before so my config cant be wrong.... reboot, error gone.
edit  sudo nano /etc/searx/settings.yml and change values to the values from ~/searx/searx/settings.yml
and wuaalala working. nice bro ty for this road trip. i just needed to change 5 files manuelly rly nice documentation helped me after reading 10 minutes and pressed at least 20 keys so continue.

please dont take this as an insult. its just sad for such a work around.
i had to ignore many errors and change many files who i had to look up all in separate sub pages from your documentation.
anyway im glad its done. :D

> 
Actually I'm quite pissed! Port 3000 still used by proxy lucky 8888 was free and i changed another service but truly trash configs in this project!

**Comment by return42** [comment]
> Actually I'm quite pissed.
> SEARX_INTERNAL_HTTP="${SEARX_INTERNAL_HTTP:-127.0.0.1:8889}"
> really messy and still not working.
> get your sh1t together instead of pointing 
> No one has the time to waste hours of reading documentation subpages without knowing where it goes...
> who the hell uses snap crap?

Is this the tone you use to expect help from the community?  We are not paid by you for anything, we are not paid by anyone for anything we do here.

Please consider that your scenario is just one you can use searx.  You can install and use searx in many ways.  Our documentation is compact and addresses various scenarios. Anyway there is much space left to improve and productive criticism is welcome.

You can choose the simple installation method using docker or the installation scripts and as admin you should have some base knowledge of services and shell scripting.

```
server:
    port : 8889
    bind_address : "127.0.0.1" # address to listen on
```

port & bind_address: see https://searx.github.io/searx/admin/settings.html

    Port number and bind address of the searx web application if you run it directly using python searx/webapp.py. Doesn’t apply to searx running on Apache or Nginx.

> .config.sh? where?
> ~/searx/utils$ ls -la | grep .config.sh
> ~/searx$ ls -la | grep .config.sh
> -rw-r--r-- 1 bla bla Dec 11 22:35 .config.sh
> nice hidden file and hidden in the sub sub documentation for https://searx.github.io/searx/utils/index.html#tooling-box-setup
> while it should be called out in: https://searx.github.io/searx/admin/installation.html#installation-scripts

Did you see the fat "Tip" box at the bottom of https://searx.github.io/searx/admin/installation.html ? .._About script’s installation options have a look at chapter Tooling box setup._  .. really what more can I do for you until you get it?

Instead you modifying searx.sh:

    SEARX_INTERNAL_HTTP="${SEARX_INTERNAL_HTTP:-127.0.0.1:8889}"

Why not doing what is written .. in .config.sh set

You set 

SEARX_INTERNAL_HTTP="127.0.0.1:8889"

and it worked for you, so what is the problem?  And if you set:

    SEARX_INTERNAL_URL="127.0.0.1:8889"

> Port 3000 still used by proxy lucky 8888 

in the same .config file

```
MORTY_LISTEN="127.0.0.1:3001"
```

> who the hell uses snap crap?

Who the hell asked you to use it?  It is intended for developer tasks, see [motivation](https://searx.github.io/searx/blog/lxcdev-202006.html#motivation) 

> or just use an install script with one default config file and than not million key pressed to get an installation done!!!!!!!!!!!!!!

The installation scripts teach you about all the step needed to install services, asking you e.g. decide to install apache or nginx .. for batch usage set environment `FORCE_TIMEOUT=0`

Clear up your mess and please consider that possibly not only we are wrong here if something went wrong in your scenario.

Over all I have to say, maintaining a searx instance is a lot of ongoing work,  if you are not willing to take it, it might be better for you to use one of the public instances: https://searx.space/

**Reply by ckorder** [reply]
nice time waste responding to every aspect of a rant instead of getting the main argument.
this rant was related to the time i wasted trying to run the great service on a non default port.
cleaning up the config files for easier deployment to avoid services port overlap would be a huge improvement. 

> Is this the tone you use to expect help from the community

what kind of help? i provided a solution to change one port.
so take the improvement and either try to implement it or ignore it.

>  We are not paid by you for anything, we are not paid by anyone for anything we do here.

Most FOSS projects get abandoned or being more trashy because there is no funding.
I'm willing to donate for the project because the idea and work done is worth it.
I currently do not have the time to dive into the project so I'm not sorry for just creating the neutral enhancement proposal for enabling easy deployment on non default port. 

And I'm still willing to donate for this issue being fixed, i would even donate for you telling me haha I don't do anything. because I'm already thankful for the work done and i could fix it by myself.

>  You can install and use searx in many ways. Our documentation is compact and addresses various scenarios. Anyway there is much space left to improve and productive criticism is welcome.

Make it easy to install. 

If its so ez to change than go ahead or silence. Please don't respond more sh1t. Thanks

**Reply by return42** [reply]
> I currently do not have the time to dive into the project 

Again: searx is not an app, it is a complex mesh up of services, it takes a lot of time and a minimal degree of cognitive performance to maintain a searx instance.

**Comment by dalf** [comment]
> nice time waste responding to every aspect of a rant instead of getting the main argument.

You ask a small team you don't know to dealing with your rant, this is not the right place, this is not the right way to ask for help / suggest changes / discuss.
